### ISSUE TITLE
Cython dependency has changed in some way?

### ISSUE THREAD

**Issue by nedbat** [gevent/gevent#757]
I don't know if this is a gevent issue or not.  On AppVeyor (a Windows CI service), I used to be able to install coverage.py and its gevent dependency just fine (ignore 2.6 for the moment): https://ci.appveyor.com/project/nedbat/coveragepy/build/default-280

Then recently (in the last week), the builds started failing for everything, because Cython wasn't available: https://ci.appveyor.com/project/nedbat/coveragepy/build/default-285

It looks like gevent requires Cython at install time, but only declares that in certain circumstances.  I don't see how that has changed recently, but something has.

Can you help?

**Comment by jamadden** [comment]
Interesting. Your failing builds seem to have `pip` grabbing `gevent-1.1rc5.tar.gz` for some reason. Yet all of the platforms doing that have a pre-built wheel available (also built on AppVeyor). Has the version of pip or `setuptools` changed recently such that the wheels might not be recognized? I know there's been a great deal of turmoil with the two of them in the last week...

As far as gevent itself goes, there have been no changes to `setup.py` since 1.1rc2 (December of last year), and they were unrelated.

I just checked and the file `gevent-1.1rc5.tar.gz` as distributed on PyPI does have the appropriate Cython-generated .c files, and they do have modification times newer than the corresponding .pyx source...However, it turns out that on Windows, gevent is not really meant to be installed from the tarball without having Cython available (our detection of when/if to try to run cython is *badly* broken on win32 and always has been as far as I can tell). That is, it is expected that Windows users will be installing the wheels (or using cygwin, possibly---wheels are new to 1.1, so cygwin or manually running cython was probably the only thing for 1.0).

Currently gevent *never* really specifies a Cython dependency in a way that matters, because the `make` step that uses cython is run **before** any `setuptools` code gets to run!

So, TLDR, it looks like some issue finding/installing the wheels. Perhaps specifying pip's `-v` option will yield more info?

**Comment by jamadden** [comment]
So on Windows cPython 3.5 with pip 7.1.2 and setuptools 18.2, I can't install any gevent wheel past version 1.1rc3; pip finds the wheels, but reports them "incompatible with this version of python". 

The naming convention of the wheels changed. In 1.1rc3 it was `gevent-1.1rc3-cp35-none-win32.whl ` whereas with 1.1rc5 it is `gevent-1.1rc5-cp35-cp35m-win32.whl `; note the platform tag differences.

The newer wheels were built with a newer version of `wheel`, and I remember something about it changing the tags (they were broken on PyPy before the change; sigh). So either we're looking at a broken wheel *build*, or it's a broken wheel *install*. I'll try some more combos, but this isn't looking like a specifically gevent problem.

**Comment by jamadden** [comment]
I upgraded pip to 8.0.3, and setuptools to 20.2.2 and I am able to install the 1.1rc5 wheel.

**Comment by jamadden** [comment]
It looks like setuptools is pinned to version 19.5 in your project?

**Comment by jamadden** [comment]
Ok, looking at [the verbose build for 3.3](https://ci.appveyor.com/project/nedbat/coveragepy/build/job/cjqisj4e4mq354wk), we can see that the first phase downloads and installs the appropriate whee; it's using pip 8.0.3l:

```
  Found link https://pypi.python.org/packages/cp33/g/gevent/gevent-1.1rc5-cp33-cp33m-win32.whl#md5=4c3c3d762785bcec849b7822de03ca6e (from https://pypi.python.org/simple/gevent/), version: 1.1rc5
```

But then the next phase, that actually runs `tox`, creates its own environment and goes back to downloading the .tar.gz:
```
Executing: C:\Python33\Scripts\tox
py33 create: C:\projects\coveragepy\.tox\py33 
py33 installdeps: nose==1.3.7, mock==1.3.0, PyContracts==1.7.9, gevent==1.1rc5, eventlet==0.18.4, greenlet==0.4.9 
...
Collecting gevent==1.1rc5
  Downloading gevent-1.1rc5.tar.gz (1.9MB)
Collecting eventlet==0.18.4
...
```

So perhaps the tox environment is out of date?

**Comment by nedbat** [comment]
Here is another build with `-vvv`: https://ci.appveyor.com/project/nedbat/coveragepy/build/default-287/job/1yti6fvkeu7t7i4a#L221

**Comment by jamadden** [comment]
I think that's pretty much the same as what I linked to, except on 3.4 instead of 3.3: the install of gevent from appveyor.yml goes according to plan, but the tox environment screws it up.

I don't know how tox locates the versions of things it copies to its virtualenv. Perhaps it doesn't get the version of pip that was [upgraded but only installed into the user location](https://github.com/nedbat/coveragepy/blob/master/appveyor.yml#L107), instead relying on the outdated system version?

**Comment by methane** [comment]
https://tox.readthedocs.org/en/latest/changelog.html#id13

**Comment by jamadden** [comment]
I've tried various combos of pip and setuptools versions. This seems specifically to be a problem in pip 7.1.2. Using *any* pip 8.x release with setuptools 18.2, 19.5 or 20.2.2 works fine, while using pip 7.1.2 with *any* of those setuptools versions always results in downloading the tarball.

**Comment by jamadden** [comment]
@methane What should I be looking for there? It talks about the `--pre` flag, but in this case the dependency on gevent is specifically pinned to `==1.1rc5` so `--pre` is irrelevant.

**Comment by nedbat** [comment]
Thanks, I've tried updating the dependencies, and I'll have another appveyor result in an hour.

**Comment by methane** [comment]
I'm sorry, it isnt related.

**Comment by jamadden** [comment]
No dice :( tox still seems to be using the broken version of pip.

I forget exactly why appveyor.yml has to do a user upgrade of pip (gevent does it too). Was it because of the 'permission denied' error due to an executable being in use (I see that locally, though the upgrade does work)? If so, maybe once the user upgrade has taken place, then the system upgrade can be done? (On the theory that the executable changes, not just contents of sys.path...I don't know how that works on Windows.) That might git tox to use the newer pip.

Another option would be to set 'sitepackages=True' in the tox environments that need gevent access. That should allow them to use the appveyor.yml-installed version of gevent, so you could remove it from the tox configuration. Obviously that decreases isolation to some extent.

Alternately, you could try installing Cython, at least until pip gets upgraded. That has the downside of not using the pre-built wheels, and building gevent is pretty slow.

**Comment by jamadden** [comment]
It looks like [upgrading virtualenv](https://github.com/nedbat/coveragepy/commit/c9995bf09bedda8e57c457fe8500d38d02086793) may have done it! At least, the 3.3 environment installed and is running the tests.

**Comment by nedbat** [comment]
Yes, upgrading virtualenv, which vendors pip, solved the .tox issues.   

Any idea about the 2.6 build on Appveyor? :)

**Comment by jamadden** [comment]
gevent doesn't build binary dists for 2.6 on Windows because it's no longer supported by the Python team, so that's probably the problem. This is the first time their lack has been noted (again, 1.0 built no binaries for Windows so this is all new). (I still build binary dists of 2.6 on OS X because 2.6 is still distributed by the OS vendor.) 

I've checked in a change that should *probably* let 2.6 sdists build on Windows and which will be in 1.1.0, but I haven't tried to test it.

I expect gevent to be dropping 2.6 support entirely after 1.1, as many projects are doing these days.

**Comment by jamadden** [comment]
@nedbat I hope you are ok with stopping testing gevent on Windows/2.6. It does turn the build green. 

Obviously gevent is indebted to your work, as are a number of my own projects. If you feel very strongly about testing 2.6 on Windows, and since you have an example of how to do it despite support, I will do what I can to get gevent bdists available there. (Given how sensitive gevent tests are  to timing, that may or may not be a notable amount of work.)

**Comment by nedbat** [comment]
@jamadden i really appreciate the work you put in today.  I'm disabling gevent+2.6 for now, when 1.1 is released, I can try enabling it again. Thanks.

**Comment by jamadden** [comment]
It's my pleasure; it's always nice to help out somebody who's helped you out. 

If you feel ready to close this issue, and open another post-release if necessary, please do; it would help with triage. Otherwise we can just leave it open for a bit.
